File: frontend/lib/vault/escrow.ts
/**
 * Shamir Secret Sharing over GF(256) — emergency access (ADR-0021 A).
 *
 * TypeScript port of `backend/src/api/vault/escrow.py`. Same algorithm:
 * polynomial per byte, Lagrange interpolation at x=0, HMAC-SHA256
 * truncated to 8 bytes for typo detection. Same `share_to_base32` format
 * → backend Python decoder accepts client-side encoded shares и наоборот.
 *
 * Per ADR §approve note «zero-knowledge preserved»: combine + decrypt
 * happen client-side; backend никогда не видит shares.
 */

const HMAC_BYTES = 8;
const HMAC_TAG = "rehome.vault.escrow.share.hmac.v1";
const BASE32_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";

export const SECRET_BYTES = 32;

export class EscrowError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "EscrowError";
  }
}

// ---------------------------------------------------------------------------
// GF(256) arithmetic — Rijndael field.

function gfMul(a: number, b: number): number {
  let p = 0;
  for (let i = 0; i < 8; i++) {
    if (b & 1) p ^= a;
    b >>= 1;
    const carry = a & 0x80;
    a = (a << 1) & 0xff;
    if (carry) a ^= 0x1b;
  }
  return p;
}

function gfInv(a: number): number {
  if (a === 0) throw new EscrowError("Division by zero in GF(256)");
  for (let x = 1; x < 256; x++) {
    if (gfMul(a, x) === 1) return x;
  }
  throw new EscrowError("Inverse not found");
}

// ---------------------------------------------------------------------------
// HMAC suffix — typo detection.

async function shareHmac(body: Uint8Array): Promise<Uint8Array> {
  const key = await crypto.subtle.importKey(
    "raw",
    new TextEncoder().encode(HMAC_TAG),
    { name: "HMAC", hash: "SHA-256" },
    /* extractable */ false,
    ["sign"],
  );
  const sig = await crypto.subtle.sign("HMAC", key, body as BufferSource);
  return new Uint8Array(sig).slice(0, HMAC_BYTES);
}

function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a[i] ^ b[i];
  return diff === 0;
}

// ---------------------------------------------------------------------------
// split + combine

export interface SplitOptions {
  threshold?: number;
  n?: number;
}

/**
 * Split `secret` into n shares; любые `threshold` recover'ят. Default
 * 2-of-2 (Variant A).
 *
 * Each share = `[index 1B][values len(secret) B][hmac 8B]`. Indices 1..n.
 */
export async function splitSecret(
  secret: Uint8Array,
  options: SplitOptions = {},
): Promise<Uint8Array[]> {
  const threshold = options.threshold ?? 2;
  const nShares = options.n ?? threshold;
  if (threshold < 2) {
    throw new EscrowError(`Threshold must be ≥2, got ${threshold}`);
  }
  if (nShares < threshold) {
    throw new EscrowError(`n (${nShares}) must be ≥ threshold (${threshold})`);
  }
  if (nShares > 255) {
    throw new EscrowError(`n (${nShares}) max 255 (1-byte index)`);
  }

  const polynomials: number[][] = [];
  const randomCoeffs = new Uint8Array(secret.length * (threshold - 1));
  crypto.getRandomValues(randomCoeffs);
  for (let i = 0; i < secret.length; i++) {
    const coeffs = [secret[i]];
    for (let j = 0; j < threshold - 1; j++) {
      coeffs.push(randomCoeffs[i * (threshold - 1) + j]);
    }
    polynomials.push(coeffs);
  }

  const shares: Uint8Array[] = [];
  for (let idx = 1; idx <= nShares; idx++) {
    const values = new Uint8Array(secret.length);
    for (let p = 0; p < polynomials.length; p++) {
      const coeffs = polynomials[p];
      let acc = 0;
      // Horner evaluation at x=idx in reverse.
      for (let c = coeffs.length - 1; c >= 0; c--) {
        acc = gfMul(acc, idx) ^ coeffs[c];
      }
      values[p] = acc;
    }
    const body = new Uint8Array(1 + values.length);
    body[0] = idx;
    body.set(values, 1);
    const hmac = await shareHmac(body);
    const share = new Uint8Array(body.length + hmac.length);
    share.set(body);
    share.set(hmac, body.length);
    shares.push(share);
  }
  return shares;
}

function lagrangeAtZero(points: Array<[number, number]>): number {
  let result = 0;
  for (let i = 0; i < points.length; i++) {
    const [xi, yi] = points[i];
    let num = 1;
    let denom = 1;
    for (let j = 0; j < points.length; j++) {
      if (i === j) continue;
      const xj = points[j][0];
      num = gfMul(num, xj);
      denom = gfMul(denom, xi ^ xj);
    }
    const term = gfMul(yi, gfMul(num, gfInv(denom)));
    result ^= term;
  }
  return result;
}

/**
 * Combine ≥2 shares → original secret. Validates HMAC + index uniqueness
 * + length consistency. Throws EscrowError на typo / corruption.
 */
export async function combineShares(shares: Uint8Array[]): Promise<Uint8Array> {
  if (shares.length < 2) {
    throw new EscrowError("Need at least 2 shares to combine");
  }
  const parsed: Array<[number, Uint8Array]> = [];
  let expectedLen: number | null = null;
  const seenIndices = new Set<number>();
  for (const share of shares) {
    if (share.length < 1 + HMAC_BYTES + 1) {
      throw new EscrowError(`Share too short: ${share.length} bytes`);
    }
    const body = share.slice(0, share.length - HMAC_BYTES);
    const suffix = share.slice(share.length - HMAC_BYTES);
    const expected = await shareHmac(body);
    if (!constantTimeEqual(expected, suffix)) {
      throw new EscrowError("Share HMAC mismatch — typo or corruption");
    }
    const idx = body[0];
    const values = body.slice(1);
    if (idx === 0) throw new EscrowError("Share index 0 reserved");
    if (seenIndices.has(idx)) {
      throw new EscrowError(`Duplicate share index ${idx}`);
    }
    seenIndices.add(idx);
    if (expectedLen === null) {
      expectedLen = values.length;
    } else if (values.length !== expectedLen) {
      throw new EscrowError("Shares have inconsistent length");
    }
    parsed.push([idx, values]);
  }
  if (expectedLen === null) {
    throw new EscrowError("No valid shares");
  }
  const secret = new Uint8Array(expectedLen);
  for (let pos = 0; pos < expectedLen; pos++) {
    const points: Array<[number, number]> = parsed.map(([idx, vals]) => [idx, vals[pos]]);
    secret[pos] = lagrangeAtZero(points);
  }
  return secret;
}

// ---------------------------------------------------------------------------
// Base32 encoding для человеко-вводимых envelope strings.

export function shareToBase32(share: Uint8Array): string {
  // RFC 4648 base32 без padding, для совместимости с Python's
  // base64.b32encode().rstrip('=').
  let bits = 0;
  let value = 0;
  let output = "";
  for (let i = 0; i < share.length; i++) {
    value = (value << 8) | share[i];
    bits += 8;
    while (bits >= 5) {
      bits -= 5;
      output += BASE32_ALPHABET[(value >>> bits) & 0x1f];
    }
  }
  if (bits > 0) {
    output += BASE32_ALPHABET[(value << (5 - bits)) & 0x1f];
  }
  return output;
}

export function base32ToShare(s: string): Uint8Array {
  // Tolerant к whitespace / lowercase (human input).
  const cleaned = s.replace(/\s+/g, "").toUpperCase();
  if (!cleaned) throw new EscrowError("Empty base32 input");
  if (!/^[A-Z2-7]+$/.test(cleaned)) {
    throw new EscrowError("Invalid base32 share (illegal characters)");
  }
  const out: number[] = [];
  let bits = 0;
  let value = 0;
  for (const ch of cleaned) {
    const idx = BASE32_ALPHABET.indexOf(ch);
    if (idx < 0) throw new EscrowError(`Invalid base32 char: ${ch}`);
    value = (value << 5) | idx;
    bits += 5;
    if (bits >= 8) {
      bits -= 8;
      out.push((value >>> bits) & 0xff);
    }
  }
  return new Uint8Array(out);
}

File: backend/src/api/vault/escrow.py
"""Shamir Secret Sharing over GF(256) — emergency access (ADR-0021 A).

Pure-Python implementation, no external crypto deps beyond stdlib `secrets`.
Used для testing + reference; production combine + decrypt происходит
client-side в admin browser (см. ADR §approve note «zero-knowledge
preserved»).

Algorithm:
- Each byte of secret = independent SSS polynomial over GF(256).
- f_i(x) = secret_i + a_1·x + a_2·x² + ... + a_(t-1)·x^(t-1)
- Share k = (k, f_0(k), f_1(k), ..., f_n(k)) — 1 byte index + n bytes.
- Combine via Lagrange interpolation at x=0.

Threshold = required shares (2 для variant A). Generates n=threshold shares
(2-of-2; no extra holders). Format per share:
- `[index 1 byte][values n bytes][hmac 8 bytes]` где hmac предотвращает
  typo'ы при ручном вводе в emergency UI.

Refs:
- https://en.wikipedia.org/wiki/Shamir%27s_secret_sharing
- https://github.com/blockstack/secret-sharing (alternate reference impl)
"""

from __future__ import annotations

import base64
import hmac
import secrets
from hashlib import sha256
from typing import Final

# Standard AES key length для escrow_key.
SECRET_BYTES: Final = 32

# HMAC suffix length per share — typo detection.
HMAC_BYTES: Final = 8

# Domain-separation tag для HMAC of shares.
_HMAC_TAG: Final = b"rehome.vault.escrow.share.hmac.v1"


class EscrowError(ValueError):
    """Generic SSS error (corrupted share, threshold not met, etc.)."""


# ---------------------------------------------------------------------------
# GF(256) arithmetic — Rijndael field (same as AES).


def _gf_mul(a: int, b: int) -> int:
    """Multiplication в GF(2^8) с reduction polynomial 0x11b."""
    p = 0
    for _ in range(8):
        if b & 1:
            p ^= a
        b >>= 1
        carry = a & 0x80
        a = (a << 1) & 0xFF
        if carry:
            a ^= 0x1B
    return p


def _gf_inv(a: int) -> int:
    """Multiplicative inverse в GF(2^8) via brute-force search.

    n=256 lookup — cost negligible vs Shamir share count (≤5)."""
    if a == 0:
        raise EscrowError("Division by zero в GF(256)")
    for x in range(1, 256):
        if _gf_mul(a, x) == 1:
            return x
    raise EscrowError("Inverse not found")  # pragma: no cover (unreachable)


# ---------------------------------------------------------------------------
# Share split / combine


def split_secret(secret: bytes, *, threshold: int = 2, n: int | None = None) -> list[bytes]:
    """Split `secret` на n shares, любые threshold из которых reconstruct'ят.

    Default: threshold=2, n=threshold (2-of-2 для variant A). Variant B
    (3-of-5) — caller передаёт threshold=3, n=5.

    Each share: [index 1B][values len(secret) B][hmac 8B]. Index = 1..n.
    """
    if threshold < 2:
        raise EscrowError(f"Threshold must be ≥2, got {threshold}")
    n_shares = n if n is not None else threshold
    if n_shares < threshold:
        raise EscrowError(f"n ({n_shares}) must be ≥ threshold ({threshold})")
    if n_shares > 255:
        raise EscrowError(f"n ({n_shares}) max 255 (1-byte index)")

    # Per byte of secret, random polynomial degree threshold-1.
    polynomials: list[list[int]] = []
    for byte_val in secret:
        coeffs = [byte_val] + [secrets.randbelow(256) for _ in range(threshold - 1)]
        polynomials.append(coeffs)

    shares: list[bytes] = []
    for idx in range(1, n_shares + 1):
        values = bytearray()
        for coeffs in polynomials:
            # Evaluate polynomial at x=idx via Horner's method.
            acc = 0
            for c in reversed(coeffs):
                acc = _gf_mul(acc, idx) ^ c
            values.append(acc)
        body = bytes([idx]) + bytes(values)
        share_hmac = _share_hmac(body)
        shares.append(body + share_hmac)
    return shares


def combine_shares(shares: list[bytes]) -> bytes:
    """Combine ≥threshold shares → original secret.

    Validates each share's HMAC + uniqueness of indices. Raises EscrowError
    при typo'е, повторе индекса, или mismatched share length.
    """
    if len(shares) < 2:
        raise EscrowError("Need at least 2 shares to combine")
    parsed: list[tuple[int, bytes]] = []
    expected_value_len: int | None = None
    seen_indices: set[int] = set()
    for share in shares:
        if len(share) < 1 + HMAC_BYTES + 1:
            raise EscrowError(f"Share too short: {len(share)} bytes")
        body, suffix = share[:-HMAC_BYTES], share[-HMAC_BYTES:]
        expected = _share_hmac(body)
        if not hmac.compare_digest(expected, suffix):
            raise EscrowError("Share HMAC mismatch — typo or corruption")
        idx = body[0]
        values = body[1:]
        if idx == 0:
            raise EscrowError("Share index 0 reserved")
        if idx in seen_indices:
            raise EscrowError(f"Duplicate share index {idx}")
        seen_indices.add(idx)
        if expected_value_len is None:
            expected_value_len = len(values)
        elif len(values) != expected_value_len:
            raise EscrowError("Shares have inconsistent length")
        parsed.append((idx, values))

    assert expected_value_len is not None
    secret = bytearray()
    for byte_pos in range(expected_value_len):
        secret.append(_lagrange_at_zero([(idx, vals[byte_pos]) for idx, vals in parsed]))
    return bytes(secret)


def _lagrange_at_zero(points: list[tuple[int, int]]) -> int:
    """Lagrange interpolation at x=0: f(0) = Σ y_i · Π (x_j / (x_j - x_i))."""
    result = 0
    for i, (xi, yi) in enumerate(points):
        # L_i(0) = Π over j≠i: (0 - x_j) / (x_i - x_j) = Π: x_j / (x_i - x_j)
        # In GF(2^8), subtraction == XOR; same as addition.
        num = 1
        denom = 1
        for j, (xj, _) in enumerate(points):
            if i == j:
                continue
            num = _gf_mul(num, xj)
            denom = _gf_mul(denom, xi ^ xj)
        term = _gf_mul(yi, _gf_mul(num, _gf_inv(denom)))
        result ^= term
    return result


def _share_hmac(body: bytes) -> bytes:
    """Truncated HMAC-SHA256 — typo detection only, not crypto auth.

    Key = domain-separation tag (public); это integrity, не secrecy.
    """
    full = hmac.new(_HMAC_TAG, body, sha256).digest()
    return full[:HMAC_BYTES]


def share_to_base32(share: bytes) -> str:
    """Encode share как human-typable base32 string (для печати на envelope).

    Base32 — case-insensitive, no ambiguous chars (0/O, 1/I) при ручном
    наборе. Length = ⌈len(share) * 8 / 5⌉; для 32-byte secret share:
    1 (idx) + 32 (vals) + 8 (hmac) = 41 bytes → 66 base32 chars (с padding).
    """
    return base64.b32encode(share).decode("ascii").rstrip("=")


def base32_to_share(s: str) -> bytes:
    """Inverse of `share_to_base32`. Tolerant к пробелам / case (human input)."""
    cleaned = "".join(s.upper().split())
    padding = "=" * (-len(cleaned) % 8)
    try:
        return base64.b32decode(cleaned + padding, casefold=True)
    except Exception as exc:
        raise EscrowError(f"Invalid base32 share: {exc}") from exc


__all__ = [
    "HMAC_BYTES",
    "SECRET_BYTES",
    "EscrowError",
    "base32_to_share",
    "combine_shares",
    "share_to_base32",
    "split_secret",
]
